### DISCUSSION TITLE
Way to get the non-zero elements from the low rank approximations of the BEM matrices.

### DISCUSSION THREAD

**Discussion by KapilKhanal** [capytaine/capytaine#598] [General]
Is there a way to just get non-zero elements ( I am assuming the HMatrix engine creates sparse matrices)?

**Comment by mancellin** [comment]
The matrices are not sparse in the sense that they contain a lot of zero. They are "data-sparse" in the sense that a few parameters are sufficient to deduce all coefficients. See https://www.mdpi.com/1996-1073/17/2/372 for more details.

The prototype implementation of H-Matrices in Capytaine is a hierarchical structure with `BlockMatrix` objects containing either other `BlockMatrix` objects or `LowRankMatrix` objects or Numpy arrays.

You can access the blocks within a `BlockMatrix` with `all_blocks`. There are typically 2x2 of them.
The `LowRankMatrix` is stored as  a `left_matrix` and a `right_matrix`.

Use `np.array` to convert a `BlockMatrix` or a `LowRankMatrix` to a Numpy array.

```python
import numpy as np
import capytaine as cpt

body = cpt.FloatingBody(mesh=cpt.mesh_parallelepiped(size=(6, 2, 2), faces_max_radius=0.08).immersed_part())
minced_body = body.minced((4, 2, 1))
gf = cpt.Delhommeau()
engine = cpt.HierarchicalToeplitzMatrixEngine(ACA_distance=1.0, ACA_tol=1e-2)
wavenumber = 1.0

S, K = engine.build_matrices(minced_body.mesh, minced_body.mesh, 0.0, np.inf, wavenumber, gf)

print(S.__class__)
# <class 'capytaine.matrices.block.BlockMatrix'>

print(S.all_blocks)
# [[<capytaine.matrices.block.BlockMatrix object at 0x77bb9f5dd780>
#   <capytaine.matrices.block.BlockMatrix object at 0x77bb9f5dc6d0>]
#  [<capytaine.matrices.block.BlockMatrix object at 0x77bb9f5dc490>
#   <capytaine.matrices.block.BlockMatrix object at 0x77bb7635fd30>]]

dense_block_example = S.all_blocks[0, 0].all_blocks[0, 0].all_blocks[0, 0]
# array([[-0.03033878-0.00093345j, -0.00834421-0.00104313j,
#         -0.00363971-0.00116575j, ..., -0.00057883-0.00222554j,
#           ...]])

lowrank_block_example = S.all_blocks[1, 0].all_blocks[1, 0].all_blocks[1, 0]
# <capytaine.matrices.low_rank.LowRankMatrix at 0x77bb763dba00>

print(lowrank_block_example.shape)
# (333, 324)
print(lowrank_block_example.left_matrix.shape)
# (333, 4)
print(lowrank_block_example.right_matrix.shape)
# (4, 324)

assert np.allclose(np.array(lowrank_block_example), lowrank_block_example.left_matrix @ lowrank_block_example.right_matrix)

```

**Reply by mancellin** [reply]
Note also that this inefficient and undocumented H-matrices prototype implementation might be extracted from Capytaine in the near future and moved into another library.

**Reply by KapilKhanal** [reply]
Gotcha, thank you @mancellin
